File: src/main/problems/java/Nr208.java
package main.problems.java;

public class Nr208 {

    class Trie {

        private final TrieNode start = new TrieNode();

        /**
         * Initialize your data structure here.
         */
        public Trie() {

        }

        /**
         * Inserts a word into the trie.
         */
        public void insert(String word) {
            TrieNode last = start;
            for (int i = 0; i < word.length(); i++) {
                char c = word.charAt(i);
                if (last.children[c - 'a'] != null) {
                    last = last.children[c - 'a'];
                } else {
                    TrieNode newTrienode = new TrieNode();
                    last.children[c - 'a'] = newTrienode;
                    last = newTrienode;
                }
            }
            last.isEnd = true;
        }

        /**
         * Returns if the word is in the trie.
         */
        public boolean search(String word) {
            TrieNode last = walkTrie(word);
            return last != null && last.isEnd;
        }

        /**
         * Returns if there is any word in the trie that starts with the given prefix.
         */
        public boolean startsWith(String prefix) {
            return walkTrie(prefix) != null;
        }

        private TrieNode walkTrie(String word) {
            TrieNode last = start;
            for (int i = 0; i < word.length(); i++) {
                char c = word.charAt(i);
                if (last.children[c - 'a'] == null) {
                    return null;
                }
                last = last.children[c - 'a'];
            }
            return last;
        }

        private class TrieNode {
            public final TrieNode[] children = new TrieNode[26];
            public boolean isEnd;
        }
    }
}

File: src/main/problems/ruby/Nr208.rb
class Trie

    class TrieNode

        attr_reader :chars

        def initialize()
            @chars = {}
        end

        def add_char(char)
            @chars[char] ||= TrieNode.new
        end

        def get_char(char)
            return @chars.values if char == '.'

            [@chars[char]]
        end

        def is_end?
            !!@chars['*']
        end
    end

    def initialize()
        @root = TrieNode.new
    end

    def insert(word)
        current = @root
        word.each_char do |char|
            current = current.add_char(char)
        end
        current.add_char('*')
    end

    def search(word)
        end_nodes = get_ending_nodes(word)
        end_nodes.any? { |node| node.is_end? }
    end

    def starts_with(prefix)
        end_nodes = get_ending_nodes(prefix)
        end_nodes.size > 0
    end

    private

    def get_ending_nodes(prefix)
        current = [@root]
        prefix.each_char do |char|
            current = current.map do |node|
                node.get_char(char)
            end.flatten.compact
        end
        current
    end
end